proposed/non-primitive-constant-data.md

```diff
@@ -0,0 +1,163 @@
+# Non-primitive Constant Data
+
+## Introduction
+Current implementations of languages in .NET are limited to a small set of possible constant types. It would be useful for computational scenarios to provide a mechanism to specify efficiently accessible pre-computed data in more complex data structures. This proposal describes a mechanism for supporting user data structures of more arbitrary data structures, supporting both array production, and constant data access to a single constant as well as a ReadOnlySpan of data.
+
+## Current valid constant forms
+Currently accepted are 4 forms of constant
+- Integer/Float constants
+```csharp
+void Constants()
+{
+   int x = 3;
+   double y = 4.0;
+}
+```
+
+Which is represented in IL via the various ldc opcodes.
+```
+IL_0001: ldc.i4.s 3
+IL_0003: stloc.s 0
+IL_0005: ldc.r8 4.0
+IL_000E: stloc.s 1
+```
+
+- String constants
+```csharp
+void Constants()
+{
+    string z = "z";
+}
+```
+
+Which is represented in IL using the ldstr instruction and #US metadata table.
+```
+IL_0001: ldstr "z"
+```
+- Byte data constants
+```csharp
+void Constants()
+{
+    ReadOnlySpan<byte> byte_data = new ReadyOnlySpan<byte>(new byte[]{0,1,2,3});
+}
+```
+Which is represented in IL using an RVA static field on an anonymous type and the accessed via ldsflda to get its address.
+```
+IL_0001: ldloca.s 0
+IL_0003: ldsflda valuetype '<PrivateImplementationDetails>'/'__StaticArrayInitTypeSize=3' '<PrivateImplementationDetails>'::'0C7A623FD2BBC05B06423BE359E4021D36E721AD'
+IL_0008: ldc.i4.3
+IL_0009: call instance void valuetype [System.Memory]System.ReadOnlySpan`1<uint8>::.ctor(void*, int32)
+```
+
+
+- Primitive data for a dynamically constructed array
+```csharp
+void Constants()
+{
+    int[] data = new byte[]{0,1,2,4};
+}
+```
+
+Which is represented in IL using an RVA static field on an anonymous type which is used in conjunction with the System.Runtime.CompilerServices.RuntimeHelpers.InitializeArray function to dynamically create an array based on a set of data which was statically placed into the assembly.
+```
+IL_0001: ldc.i4.4
+IL_0002: newarr [mscorlib]System.Byte
+IL_0007: dup
+IL_0008: ldtoken field int32 '<PrivateImplementationDetails>'::'12DADA1FFF4D4787ADE3333147202C3B443E376F'
+IL_000d: call void [mscorlib]System.Runtime.CompilerServices.RuntimeHelpers::InitializeArray(class [mscorlib]System.Array, valuetype [mscorlib]System.RuntimeFieldHandle)
+IL_0012: stloc.0
+```
+## New valid constant forms
+
+The proposal is to provide a set of CompilerServices apis implemented as intrinsics (in most cases) which will allow the use of more complex constants. This api will accept as a byref argument a reference to data in a well defined type layout, and return a byref to data in the correct platform layout.
+
+### Definition of types which can be represented as constants
+- Must be a struct
```

Comment:
Can it be a generic type? - like `int?` ?
It feels "why not", but then we must be sure that such types respect sequential layout and have predictable packing.

Reply:
Well, `int?` is probably a bad example, since it has private fields, but the question still remains - would generic types be supported?
A better example would be tuples like  `(byte, byte, byte)`  - can that be a constant?

Reply:
I think this could be a generic type, but it only make sense if we were willing to make setting the fields of a nullable a well known detail of the frameworks. Given the history of nullable<T> I think it would be fine, but it would probably require special casing in the runtime to have this knowledge.

Reply:
And yes, it would be ok to declare the fields of ValueTuple<T> to be a public contract as well. With regards to layout of generic structures, by allowing the layout to differ between the PE file and the actual runtime, support for a variety of capabilities pops out. I hadn't thought of generics, but they should just work.